### PR TITLE
master lpd 27538

### PR DESCRIPTION
**Pull request by kyle-miho**
https://liferay.atlassian.net/browse/LPD-27538
cc: @drewbrokke @nicolas-sss

### CODE REVIEW

**Comment by kyle-miho** [comment]
ci:test:sf

**Review by drewbrokke — APPROVED** [review]

**Comment by kyle-miho** [comment]
ci:test:hotel

**Comment by kyle-miho** [comment]
ci:test:relevant

**Review by nicolas-sss — APPROVED** [review]

File: modules/util/source-formatter/src/main/resources/dependencies/replacements.json [hunk 1]
```diff
@@ -21,6 +21,11 @@
 		],
 		"to": "getFriendlyURL(new InfoItemReference(param#0#, param#1#), param#2#)"
 	},
+	{
+		"from": "JSONUtil.isValid(String paramName)",
+		"issueKey": "LPD-8007",
+		"to": "JSONUtil.isJSONObject(param#0#)"
+	},
```
Comment (nicolas-sss) [inline modules/util/source-formatter/src/main/resources/dependencies/replacements.json:28]: hi @kyle-miho when the parameter will not be changed and will only be a rename it is not necessary to add this information, and when the class will not be changed it is not added in the “`from`” attribute, it is only used for verification by being placed in “`classNames`”.  e.g.:
```
	{
		"classNames": [
			"JSONUtil"
		],
		"from": "isValid",
		"issueKey": "LPD-8007",
		"to": "isJSONObject"
	},
```
Parameter checking would be relevant, for example, if the class had more than one `isValid` method and only one had been replaced by the other.

**Review by nicolas-sss — CHANGES_REQUESTED** [review]

File: modules/util/source-formatter/src/main/resources/dependencies/replacements.json (hunk 1 above)
Comment (nicolas-sss) [reply modules/util/source-formatter/src/main/resources/dependencies/replacements.json:28]: Another Process that I would Like to suggest is:
The pattern when registered on the ticket is based on the specific Upgrade execution that was found.

Some patterns could happen on other file types and it's up to the judgment of the developers to identify these cases. 
This case, for example, being in a utility class, could appear in JSPs, JSPFs, etc. Of course, we can't guess which cases will come along and we really should worry about what we've already found, but it's interesting at least to look in the Liferay Portal code at the uses of this class and/or method. 
If they are present in other file types, it's worth evaluating the likelihood of them occurring in that other type vs. the cost of developing them.
If:
Cost is high but so is value, it's a good case to analyze.
If the cost is low and the value is high, is valid to develop. And both having a low cost, it also makes sense.

In this scenario we would have: 

```
	{
		"from": "JSONUtil.isValid",
		"issueKey": "LPD-8007",
		"to": "JSONUtil.isJSONObject",
		"validExtensions": [
			"java",
			"jsp"
		]
	},
```

**Comment by kyle-miho** [comment]
ci:test:hotel

**Comment by kyle-miho** [comment]
ci:test:sf

**Comment by kyle-miho** [comment]
@nicolas-sss I fixed the pattern, also including JSP files since I did a mass search and found that there are cases where JSONUtil is used. The specific method may not be used currently, but I agree that we should cover it in a chance the customer does use JSONUtil

**Review by nicolas-sss — CHANGES_REQUESTED** [review]
Hi @kyle-miho, good to know. But like I said before, the class name tag only is used on Java files. So it will be ignored in JSP files and catch any isValid method, at least that was the previous implementation. Could you please check? Because of that we should do like this:

```
	{
		"from": "JSONUtil.isValid",
		"issueKey": "LPD-8007",
		"to": "JSONUtil.isJSONObject",
		"validExtensions": [
			"java",
			"jsp"
		]
	},
```
This way the name of the class goes directly on the regex, specific for Utility class cases. Without this, there should be two JSON, objects if you are sure what the variable will be called on the jSP. There is an example with themeDisplay there.

**Comment by kyle-miho** [comment]
ci:test:sf

**Comment by kyle-miho** [comment]
@nicolas-sss Okay didnt realize classNames didnt support non java files. I wanted to add a check to make sure we only used classNames with java files, but apparently there is 1 other check that can use classNames in non java files?

`	{
		"classNames": [
			"Indexer"
		],
		"from": "getClassNames",
		"issueKey": "LPS-191962",
		"to": "getSearchClassNames",
		"validExtensions": [
			"java",
			"jspf"
		]
	},`

**Comment by nicolas-sss** [comment]
> @nicolas-sss Okay didnt realize classNames didnt support non java files. I wanted to add a check to make sure we only used classNames with java files, but apparently there is 1 other check that can use classNames in non java files?
> 
> `{ "classNames": [ "Indexer" ], "from": "getClassNames", "issueKey": "LPS-191962", "to": "getSearchClassNames", "validExtensions": [ "java", "jspf" ] },`
> 
> Does this seem like a valid usecase, or is it being parsed incorrectly in jspf files?

Good catch! It's the same case as yours, it's being ignored and will replace any getClassNames in `jspf` files if find it.
Could you fix it as well? To fix this case I would recommend studying the code flow for non-Java classes.

The problem is that the Indexer can be instantiated with any name on the variable and the methods used to retrieve and validate the type are made to be used on Java classes. The structure on Jspf files is a little trick and we haven't an easy way to be sure of the type of the variable before replacing the method. 
You could think of a solution or if isn't possible create another object with the `jspf` case that will be only a message case.

**Comment by kyle-miho** [comment]
@nicolas-sss Messages arent currently supported for non java files. I tried making a lot of changes to fix it and support it, but it probably isnt worth doing since it ultimately got roadblocked by needing to getClassName working for jsp/jspf files to determine if the parameters are correct. If you dont do this then it will still make replacements even with the wrong parameters (including null). Hacking around it doesnt make sense so I will probably just send this as is, and implementing Messages (I have this stashed) and getClassName for JSP/JSPF files can be slated for the future? What do you think

**Comment by kyle-miho** [comment]
ci:test:hotel

**Comment by kyle-miho** [comment]
ci:test:sf

**Comment by kyle-miho** [comment]
ci:test:sf

**Comment by kyle-miho** [comment]
https://github.com/brianchandotcom/liferay-portal/pull/152152

**Comment by kyle-miho** [comment]
Merge conflicts fixed. @nicolas-sss can you check this PR when you have a chance? Thanks!

**Comment by kyle-miho** [comment]
ci:test:sf

**Comment by kyle-miho** [comment]
ci:test:hotel

**Review by nicolas-sss — APPROVED** [review]

**Comment by kyle-miho** [comment]
https://github.com/brianchandotcom/liferay-portal/pull/152474

**Pull request closed without merging**